### ISSUE TITLE
[ENH] support `pandas` input with nullable `dtypes`

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#2967]
Update: currently, the input checkers don't let nullable types pass, an error is thrown.

I think we should move to a detect & coerce solution, with optional config to turn off the input checks/coercions and let an advanced user try their luck with the estimator...

---
A number of estimators fail when being passed `pandas` formatted input with nullable `dtypes`, e.g., `Float64` instead of `float64`.

There seems to be a general problem with nullable `dtype` passed to `sktime` estimators, I've opened a diagnostic PR here: https://github.com/alan-turing-institute/sktime/pull/2966

**Comment by fkiraly** [comment]
@tamargrey, yes, this is a known issue - what would be your preferred behaviour?

What I could imagine:

* forecast is the same if no NaN are present
* column types are preserved

**Comment by fkiraly** [comment]
I've updated the diagnostic PR to see how widespread the problem currently is, and whether it catches `ExponentialSmoothing`:
https://github.com/sktime/sktime/pull/2966

Another question, @tamargrey - is coercing to non-nullable (numpy) types a viable workaround for you? If not, why not?

**Comment by tamargrey** [comment]
@fkiraly Your suggested behavior makes sense to me, and coercing to non nullable types is our existing workaround, so we can keep working with that in the mean time. The downside to having to coerce types is, of course, the added complexity that we have to maintain.

**Comment by fkiraly** [comment]
> The downside to having to coerce types is, of course, the added complexity that we have to maintain.

We could do that at input, if it helps.
(of course, would have to check runtime impact etc first - our checks and conversions are already a bit slow).
Do you have any code snippet that does the coercion for you?

**Comment by tamargrey** [comment]
Runtimes are also a consideration for us, so I definitely agree that that should be taken into account to see if there's a better workaround. 

We currently coerce `boolean` to `bool` if nans are not present and `category` if they are and `Int64` to `float64` (we do not yet use `Float64`). 
```python
  if str(y.dtype) == "Int64":
      y = y.astype("float64")
  elif str(y.dtype) == "boolean":
      if not y.isnull().any():
          y = y.astype("bool")
      else:
          y = y.astype("category")
```

For further context, I'll link to EvalML's [ReplaceNullableTypes transformer](https://github.com/alteryx/evalml/blob/main/evalml/pipelines/components/transformers/preprocessing/replace_nullable_types.py#L78-L85), which is what does the bulk of the coercion described above in our AutoML pipelines, though it performs the coercion with our own typing library, [Woodwork](https://woodwork.alteryx.com/en/stable/), which ultimately uses the numpy and pandas dtypes but also adds other semantic information.

**Comment by fkiraly** [comment]
Hm, is the `evalml` `Transformer` `sklearn` compatible? You could just pipeline it with any `sktime` transformer, although the suger won't work due to it not directly inheriting from `sklearn` `BaseEstimator`. Also, the checks might not let the nullable types through, maybe the solution would be to turn off checks (I still want to write that config functionality)

**Comment by tamargrey** [comment]
Yeah, our current workaround involves applying the replace nullable types transformer at the beginning of all of the Pipelines EvalML generates (which aren't that dissimilar from the sklearn pipelines in concept!). 

I also found nullable type incompatibilities in the AutoARIMA regressor -  https://github.com/sktime/sktime/issues/4236. The error AutoARIMA is producing is different enough that I wanted to give it its own issue, but thought I'd link it here incase these are connected. Our current workaround is the same as for the `ExponentialSmoothing` regressor.

**Comment by fkiraly** [comment]
Given that current specifications prevent passing of nullable types, this has turned from a bug into a feature request (which I think is sensible). Updated the issue description.

**Comment by tamargrey** [comment]
@fkiraly That approach sounds good to me.

**Comment by fkiraly** [comment]
Let's see where that gets us - pinging @danbartl and @hoesler in case they might have something to add, as they have been the two people most recently working on (or close to) the checker/converter logic.

**Comment by fkiraly** [comment]
Partial work here: https://github.com/sktime/sktime/pull/4245

There's still some testing needed - what would be great is some (possibly hypothetical) example code that you think should work at the end state that I can use to test against, @tamargrey.

**Comment by tamargrey** [comment]
In the same way the following block of code with the non nullable dtypes works:
```python
    for dtype in ['int64', 'float64', 'bool']:
        X = pd.DataFrame()
        X["ints"] = pd.Series([i for i in range(40)], dtype=dtype)
        X.index = pd.date_range("1/1/21", periods=40)
        y = pd.Series([1, 0] * 20, dtype=dtype)

        sk_clf = ExponentialSmoothing()

        
        
        sk_clf.fit(X=None, y=y)


        fh_ = ForecastingHorizon(
            [i + 1 for i in range(len(X))],
            is_relative=True,
        )
        sk_clf.predict(fh=fh_, X=X)
```

We want the block of code below to work with the nullable dtypes (the only difference is the dtypes specified):

```python
    for dtype in ['Int64', 'Float64', 'boolean']:
        X = pd.DataFrame()
        X["ints"] = pd.Series([i for i in range(40)], dtype=dtype)
        X.index = pd.date_range("1/1/21", periods=40)
        y = pd.Series([1, 0] * 20, dtype=dtype)

        sk_clf = ExponentialSmoothing()

        
        
        sk_clf.fit(X=None, y=y)


        fh_ = ForecastingHorizon(
            [i + 1 for i in range(len(X))],
            is_relative=True,
        )
        sk_clf.predict(fh=fh_, X=X)
```

**Comment by tamargrey** [comment]
@fkiraly I just updated the comment above to fix a couple of typos - just an fyi in case you were trying to use the snippet and it wasn't working.

**Comment by fkiraly** [comment]
ah, I see - `boolean` should be cast to `float` too, I presume.

The other two already seem to work in https://github.com/sktime/sktime/pull/4245, but I need to add more systematic tests.

**Comment by fkiraly** [comment]
#4245 should work - the failures in the diagnostic https://github.com/sktime/sktime/pull/2966 were due to a different bug in the classifier framework (inconsistent interface for private `_get_train_probs` that someone stealth-introduced), and the unreported bug https://github.com/sktime/sktime/pull/4275

I've added the case above as a test, and we will probably add some nullable types in scenarios as test case along the lines of #2966.
What I would appreciate is if you could check whether performance is impacted in any discernible way by #4245, especially for large panel and hierarchical data sets.

**Comment by fkiraly** [comment]
I think we can consider this closed by #4245.
